File: shared/domain/src/finance.ts
// Shared retirement, market-quote, and net-worth contracts.
//
// Retirement rows are synchronized household data. Quotes are operational
// observations with independent availability and freshness. Keeping the two
// contracts separate prevents a cache observation from masquerading as a
// durable holding value.

import { type FamilyMember, netWorthScopeFor } from "./family.ts"
import {
  type Cents,
  type Sats,
  satsToUsdCents,
  sharesToValueCents,
  sum,
  usdCentsToSats,
} from "./money.ts"

export const MARKET_SYMBOLS = ["BTC", "VOO", "IBIT"] as const
export type MarketSymbol = (typeof MARKET_SYMBOLS)[number]
export type MarketQuoteStatus = "live" | "stale" | "unavailable"

/**
 * A single evidenced market observation.
 *
 * Live and stale observations carry a positive integer-cent price and fetch
 * time. Unavailable is explicit and carries no price; it is never `$0.00`.
 */
export interface MarketQuote {
  readonly symbol: MarketSymbol
  readonly priceCents: Cents | null
  readonly source: string
  readonly fetchedAt: string | null
  readonly status: MarketQuoteStatus
}

/** Structurally complete even when one or more observations are unavailable. */
export interface MarketQuoteSnapshot {
  readonly quotes: readonly MarketQuote[]
}

export function isMarketSymbol(value: unknown): value is MarketSymbol {
  return typeof value === "string" && (MARKET_SYMBOLS as readonly string[]).includes(value)
}

/** Refuse contradictory observation states at an adapter boundary. */
export function assertMarketQuote(quote: MarketQuote): MarketQuote {
  if (!isMarketSymbol(quote.symbol)) {
    throw new RangeError(`Unsupported market symbol: ${String(quote.symbol)}`)
  }
  if (quote.status !== "live" && quote.status !== "stale" && quote.status !== "unavailable") {
    throw new RangeError(`Unsupported market quote status: ${String(quote.status)}`)
  }
  if (quote.source.trim() === "") throw new RangeError("Market quote source must not be empty")

  if (quote.status === "unavailable") {
    if (quote.priceCents !== null) {
      throw new RangeError(`Unavailable ${quote.symbol} quote must not carry a price`)
    }
    return quote
  }

  if (quote.priceCents === null || quote.priceCents <= 0n) {
    throw new RangeError(`${quote.status} ${quote.symbol} quote must carry a positive price`)
  }
  if (quote.fetchedAt === null || quote.fetchedAt.trim() === "") {
    throw new RangeError(`${quote.status} ${quote.symbol} quote must carry fetchedAt`)
  }
  return quote
}

/** Require exactly one BTC, VOO, and IBIT entry. */
export function assertMarketQuoteSnapshot(snapshot: MarketQuoteSnapshot): MarketQuoteSnapshot {
  if (snapshot.quotes.length !== MARKET_SYMBOLS.length) {
    throw new RangeError(`Market quote snapshot must contain ${MARKET_SYMBOLS.length} quotes`)
  }

  const seen = new Set<MarketSymbol>()
  for (const quote of snapshot.quotes) {
    assertMarketQuote(quote)
    if (seen.has(quote.symbol)) throw new RangeError(`Duplicate market quote: ${quote.symbol}`)
    seen.add(quote.symbol)
  }
  for (const symbol of MARKET_SYMBOLS) {
    if (!seen.has(symbol)) throw new RangeError(`Missing market quote: ${symbol}`)
  }
  return snapshot
}

/** Return an observation including explicit unavailable state, or null if absent. */
export function marketQuoteFor(
  quotes: readonly MarketQuote[],
  symbol: MarketSymbol,
): MarketQuote | null {
  const quote = quotes.find((candidate) => candidate.symbol === symbol) ?? null
  return quote === null ? null : assertMarketQuote(quote)
}

/** Live and stale are usable; unavailable and absent are not. */
export function usableMarketQuote(
  quotes: readonly MarketQuote[],
  symbol: MarketSymbol,
): MarketQuote | null {
  const quote = marketQuoteFor(quotes, symbol)
  return quote?.status === "live" || quote?.status === "stale" ? quote : null
}

export interface FinanceLot {
  readonly date: string
  readonly type: string
  readonly pricePerShareCents: Cents
  readonly sharesDecimal: string
  readonly amountInvestedCents: Cents
  readonly note: string | null
}

export interface FinanceHolding {
  readonly name: string
  readonly category: string
  readonly ticker: string | null
  readonly valueCents: Cents
  readonly costBasisCents: Cents
  readonly gainBps: bigint
  readonly sharesDecimal: string
  readonly avgCostCents: Cents
  readonly currentPricePerShareCents: Cents
  readonly isProxy: boolean
  readonly proxyNote: string | null
  readonly lots: readonly FinanceLot[]
}

export interface FinanceAccount {
  readonly key: string
  readonly owner: FamilyMember
  readonly provider: string
  readonly totalValueCents: Cents
  readonly weeklyContributionCents: Cents
  /** Schedule metadata such as "Friday"; never a trade instruction. */
  readonly weeklyContributionDay: string | null
  readonly holdings: readonly FinanceHolding[]
}

export interface FinanceDocument {
  readonly updatedAtMs: number
  readonly lastUpdated: string
  /** Compatibility projection only; selectors derive scoped accounts once. */
  readonly retirementTotalCents: Cents | null
  readonly accounts: readonly FinanceAccount[]
}

export type HoldingValuationBasis = "market-quote" | "stored-value"

export interface HoldingValuation {
  readonly holding: FinanceHolding
  readonly valueCents: Cents
  readonly basis: HoldingValuationBasis
  /** Includes explicit unavailable state when the snapshot supplied it. */
  readonly quote: MarketQuote | null
}

export interface AccountValuation {
  readonly account: FinanceAccount
  readonly valueCents: Cents
  readonly holdings: readonly HoldingValuation[]
}

/**
 * Revalue VOO/IBIT from a usable quote. Other holdings keep their synchronized
 * stored value. Stale observations are usable but remain labeled stale.
 */
export function valueFinanceHolding(
  holding: FinanceHolding,
  quotes: readonly MarketQuote[],
): HoldingValuation {
  const ticker = holding.ticker?.trim().toUpperCase() ?? ""
  const symbol: MarketSymbol | null = ticker === "VOO" || ticker === "IBIT" ? ticker : null
  const observation = symbol === null ? null : marketQuoteFor(quotes, symbol)
  const quote = observation?.status === "live" || observation?.status === "stale"
    ? observation
    : null

  if (quote !== null) {
    const priceCents = quote.priceCents
    if (priceCents === null) throw new RangeError(`${quote.symbol} quote lost its validated price`)
    return {
      holding,
      valueCents: sharesToValueCents(holding.sharesDecimal, priceCents),
      basis: "market-quote",
      quote,
    }
  }

  return {
    holding,
    valueCents: holding.valueCents,
    basis: "stored-value",
    quote: observation,
  }
}

/** Holdings are detail when their sum is positive; account total is fallback-only. */
export function valueFinanceAccount(
  account: FinanceAccount,
  quotes: readonly MarketQuote[],
): AccountValuation {
  const holdings = account.holdings.map((holding) => valueFinanceHolding(holding, quotes))
  const holdingTotal = sum(holdings.map((holding) => holding.valueCents))
  return {
    account,
    valueCents: holdingTotal > 0n ? holdingTotal : account.totalValueCents,
    holdings,
  }
}

export interface NetWorthSelection {
  /** Canonical already-net-worth-scoped Bitcoin balance supplied by the adapter. */
  readonly bitcoinSats: Sats
  readonly bitcoinValueCents: Cents | null
  readonly retirementValueCents: Cents
  readonly retirementValueSats: Sats | null
  readonly totalValueCents: Cents | null
  readonly totalValueSats: Sats | null
  readonly accounts: readonly AccountValuation[]
  /** Preserves live, stale, unavailable, or absent quote state. */
  readonly btcQuote: MarketQuote | null
}

/**
 * Adult household net worth is eligible Bitcoin plus adult retirement once.
 *
 * The adapter must supply the canonical Bitcoin balance already scoped for net
 * worth. Retirement accounts are defense-in-depth scoped here, excluding child
 * accounts from adult totals even though adults may see those accounts. The
 * document-level retirement projection is intentionally not an input.
 */
export function selectNetWorth(input: {
  readonly viewer: FamilyMember
  readonly bitcoinSats: Sats
  readonly financeAccounts: readonly FinanceAccount[]
  readonly quotes: readonly MarketQuote[]
}): NetWorthSelection {
  const accounts = netWorthScopeFor(input.viewer, input.financeAccounts)
    .map((account) => valueFinanceAccount(account, input.quotes))
  const retirementValueCents = sum(accounts.map((account) => account.valueCents))
  const btcObservation = marketQuoteFor(input.quotes, "BTC")
  const btcQuote = btcObservation?.status === "live" || btcObservation?.status === "stale"
    ? btcObservation
    : null

  if (btcQuote === null) {
    return {
      bitcoinSats: input.bitcoinSats,
      bitcoinValueCents: null,
      retirementValueCents,
      retirementValueSats: null,
      totalValueCents: null,
      totalValueSats: null,
      accounts,
      btcQuote: btcObservation,
    }
  }

  const btcPriceCents = btcQuote.priceCents
  if (btcPriceCents === null) throw new RangeError("BTC quote lost its validated price")
  const bitcoinValueCents = satsToUsdCents(input.bitcoinSats, btcPriceCents)
  const retirementValueSats = usdCentsToSats(retirementValueCents, btcPriceCents)
  return {
    bitcoinSats: input.bitcoinSats,
    bitcoinValueCents,
    retirementValueCents,
    retirementValueSats,
    totalValueCents: bitcoinValueCents + retirementValueCents,
    totalValueSats: input.bitcoinSats + retirementValueSats,
    accounts,
    btcQuote,
  }
}

File: android/domain/src/main/kotlin/com/sats21m/vogelvault/domain/Finance.kt
package com.sats21m.vogelvault.domain

import java.math.BigInteger
import java.time.Instant

/**
 * Synced retirement data and operational quotes intentionally use different
 * contracts. A quote cached on one device is not a Convex retirement record.
 */
enum class MarketSymbol { BTC, VOO, IBIT }

enum class MarketQuoteStatus { LIVE, STALE, UNAVAILABLE }

data class MarketQuote(
    val symbol: MarketSymbol,
    val priceCents: Long?,
    val source: String,
    val fetchedAt: String?,
    val status: MarketQuoteStatus,
) {
    init {
        require(source.isNotBlank()) { "Market quote source must not be empty" }
        when (status) {
            MarketQuoteStatus.UNAVAILABLE ->
                require(priceCents == null) { "Unavailable $symbol quote must not carry a price" }
            MarketQuoteStatus.LIVE, MarketQuoteStatus.STALE -> {
                require(priceCents != null && priceCents > 0L) {
                    "$status $symbol quote must carry a positive price"
                }
                require(fetchedAt != null && fetchedAt.isCanonicalQuoteInstant()) {
                    "$status $symbol quote must carry a canonical ISO-8601 fetchedAt"
                }
            }
        }
    }

    val isUsable: Boolean
        get() = status != MarketQuoteStatus.UNAVAILABLE
}

private val canonicalQuoteInstant =
    Regex("""\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{3})?Z""")

private fun String.isCanonicalQuoteInstant(): Boolean {
    if (!canonicalQuoteInstant.matches(this)) return false
    val normalized = try {
        Instant.parse(this).toString()
    } catch (_: java.time.format.DateTimeParseException) {
        return false
    }
    return this == normalized || this == normalized.removeSuffix("Z") + ".000Z"
}

data class MarketQuoteSnapshot(val quotes: List<MarketQuote>) {
    init {
        require(quotes.size == MarketSymbol.entries.size) {
            "Market quote snapshot must contain ${MarketSymbol.entries.size} quotes"
        }
        require(quotes.map { it.symbol }.toSet() == MarketSymbol.entries.toSet()) {
            "Market quote snapshot must contain BTC, VOO, and IBIT exactly once"
        }
    }
}

fun List<MarketQuote>.marketQuoteFor(symbol: MarketSymbol): MarketQuote? =
    firstOrNull { it.symbol == symbol }

fun List<MarketQuote>.usableQuote(symbol: MarketSymbol): MarketQuote? =
    marketQuoteFor(symbol)?.takeIf { it.isUsable }

data class FinanceLot(
    val date: String,
    val type: String,
    val pricePerShareCents: Long,
    val sharesDecimal: String,
    val amountInvestedCents: Long,
    val note: String? = null,
)

data class FinanceHolding(
    val name: String,
    val category: String,
    val ticker: String?,
    val valueCents: Long,
    val costBasisCents: Long,
    val gainBps: Long,
    val sharesDecimal: String,
    val avgCostCents: Long,
    val currentPricePerShareCents: Long,
    val isProxy: Boolean,
    val proxyNote: String? = null,
    val lots: List<FinanceLot> = emptyList(),
)

data class FinanceAccount(
    val key: String,
    override val owner: FamilyMember,
    val provider: String,
    val totalValueCents: Long,
    val weeklyContributionCents: Long,
    /** Synced schedule metadata, not an instruction to execute a market order. */
    val weeklyContributionDay: String?,
    val holdings: List<FinanceHolding>,
) : Owned

data class FinanceDocument(
    val updatedAtMs: Long,
    val lastUpdated: String,
    /**
     * Compatibility projection only. Net worth derives scoped [accounts] and
     * never adds this total again.
     */
    val retirementTotalCents: Long?,
    val accounts: List<FinanceAccount>,
)

enum class HoldingValuationBasis { MARKET_QUOTE, STORED_VALUE }

data class HoldingValuation(
    val holding: FinanceHolding,
    val valueCents: Long,
    val basis: HoldingValuationBasis,
    val quote: MarketQuote?,
)

data class AccountValuation(
    val account: FinanceAccount,
    val valueCents: Long,
    val holdings: List<HoldingValuation>,
)

fun FinanceHolding.marketValue(quotes: List<MarketQuote>): HoldingValuation {
    val symbol = when (ticker?.trim()?.uppercase()) {
        "VOO" -> MarketSymbol.VOO
        "IBIT" -> MarketSymbol.IBIT
        else -> null
    }
    val observation = symbol?.let(quotes::marketQuoteFor)
    val quote = observation?.takeIf { it.isUsable }
    return if (quote != null) {
        HoldingValuation(
            holding = this,
            valueCents = Money.sharesToValueCents(sharesDecimal, checkNotNull(quote.priceCents)),
            basis = HoldingValuationBasis.MARKET_QUOTE,
            quote = quote,
        )
    } else {
        HoldingValuation(this, valueCents, HoldingValuationBasis.STORED_VALUE, observation)
    }
}

fun FinanceAccount.marketValue(quotes: List<MarketQuote>): AccountValuation {
    val holdingValues = holdings.map { it.marketValue(quotes) }
    val holdingTotal = holdingValues.fold(0L) { total, holding ->
        Math.addExact(total, holding.valueCents)
    }
    return AccountValuation(
        account = this,
        valueCents = if (holdingTotal > 0L) holdingTotal else totalValueCents,
        holdings = holdingValues,
    )
}

data class NetWorthSelection(
    val bitcoinSats: Long,
    val bitcoinValueCents: Long?,
    val retirementValueCents: Long,
    val retirementValueSats: Long?,
    val totalValueCents: Long?,
    val totalValueSats: Long?,
    val accounts: List<AccountValuation>,
    val btcQuote: MarketQuote?,
)

/**
 * Select one profile's net worth without adding the same retirement holdings
 * both as account values and as the document-level retirement projection.
 */
fun selectNetWorth(
    viewer: FamilyMember,
    bitcoinSats: Long,
    financeAccounts: List<FinanceAccount>,
    quotes: List<MarketQuote>,
): NetWorthSelection {
    val accounts = financeAccounts.netWorthScopeFor(viewer).map { it.marketValue(quotes) }
    val retirementCents = accounts.fold(0L) { total, account ->
        Math.addExact(total, account.valueCents)
    }
    val btcObservation = quotes.marketQuoteFor(MarketSymbol.BTC)
    val btcQuote = btcObservation?.takeIf { it.isUsable }
        ?: return NetWorthSelection(
            bitcoinSats = bitcoinSats,
            bitcoinValueCents = null,
            retirementValueCents = retirementCents,
            retirementValueSats = null,
            totalValueCents = null,
            totalValueSats = null,
            accounts = accounts,
            btcQuote = btcObservation,
        )

    val price = checkNotNull(btcQuote.priceCents)
    val bitcoinCents = Money.satsToUsdCents(bitcoinSats, price)
    val retirementSats = Money.usdCentsToSats(retirementCents, price)
    return NetWorthSelection(
        bitcoinSats = bitcoinSats,
        bitcoinValueCents = bitcoinCents,
        retirementValueCents = retirementCents,
        retirementValueSats = retirementSats,
        totalValueCents = Math.addExact(bitcoinCents, retirementCents),
        totalValueSats = Math.addExact(bitcoinSats, retirementSats),
        accounts = accounts,
        btcQuote = btcQuote,
    )
}

enum class BudgetHealthStatus { ON_TRACK, CLOSE, OVER }

data class BudgetHealth(
    val status: BudgetHealthStatus,
    val label: String,
    val usedPercent: Int,
    val barBasisPoints: Int,
    val remainingPercent: Int?,
    val overPercent: Int?,
)

/**
 * iOS parity: green below 85%, yellow from 85% through exactly 100%, red only
 * when spend exceeds the limit. Percentages are truncated whole numbers.
 */
fun budgetHealth(plannedCents: Long, spentCents: Long): BudgetHealth {
    if (plannedCents <= 0L) {
        val isOver = spentCents > plannedCents
        return BudgetHealth(
            status = if (isOver) BudgetHealthStatus.OVER else BudgetHealthStatus.ON_TRACK,
            label = if (isOver) "OVER" else "ON TRACK",
            usedPercent = if (isOver) 200 else 0,
            barBasisPoints = if (isOver) 10_000 else 0,
            remainingPercent = if (isOver) null else 100,
            overPercent = if (isOver) 100 else null,
        )
    }

    val planned = BigInteger.valueOf(plannedCents)
    val spent = BigInteger.valueOf(spentCents)
    val rawBps = spent.multiply(BigInteger.valueOf(10_000L)).divide(planned)
    val visualBps = rawBps.coerceIn(BigInteger.ZERO, BigInteger.valueOf(20_000L))
    val isOver = spentCents > plannedCents
    val isClose = rawBps >= BigInteger.valueOf(8_500L) && !isOver
    val status = when {
        isOver -> BudgetHealthStatus.OVER
        isClose -> BudgetHealthStatus.CLOSE
        else -> BudgetHealthStatus.ON_TRACK
    }
    val remaining = if (isOver) null else {
        planned.subtract(spent).multiply(BigInteger.valueOf(100L)).divide(planned).toInt()
    }
    val over = if (isOver) {
        spent.subtract(planned)
            .multiply(BigInteger.valueOf(100L))
            .divide(planned)
            .coerceIn(BigInteger.ZERO, BigInteger.valueOf(100L))
            .toInt()
    } else {
        null
    }
    return BudgetHealth(
        status = status,
        label = when (status) {
            BudgetHealthStatus.ON_TRACK -> "ON TRACK"
            BudgetHealthStatus.CLOSE -> "CLOSE"
            BudgetHealthStatus.OVER -> "OVER"
        },
        usedPercent = visualBps.divide(BigInteger.valueOf(100L)).toInt(),
        barBasisPoints = visualBps.coerceAtMost(BigInteger.valueOf(10_000L)).toInt(),
        remainingPercent = remaining,
        overPercent = over,
    )
}

/**
 * Exact category rows for one viewer and month, including refunds/credits that
 * reduce the signed category spend.
 */
fun List<Transaction>.budgetCategoryTransactionsFor(
    viewer: FamilyMember,
    month: String,
    category: String,
): List<Transaction> =
    budgetTransactionsFor(viewer)
        .inMonth(month)
        .filter { it.category == category }
